uamqp/error.py

```diff
@@ -7,10 +7,138 @@
+    FrameSizeTooSmall = b"amqp:frame-size-too-small"
+    ResourceLimitExceeded = b"amqp:resource-limit-exceeded"
+    PreconditionFailed = b"amqp:precondition-failed"
+    ConnectionRedirect = b"amqp:connection:redirect"
+    ConnectionCloseForced = b"amqp:connection:forced"
+    ConnectionFramingError = b"amqp:connection:framing-error"
+    SessionWindowViolation = b"amqp:session:window-violation"
+    SessionErrantLink = b"amqp:session:errant-link"
+    SessionHandleInUse = b"amqp:session:handle-in-use"
+    SessionUnattachedHandle = b"amqp:session:unattached-handle"
+    LinkRedirect = b"amqp:link:redirect"
+    LinkStolen = b"amqp:link:stolen"
+    LinkDetachForced = b"amqp:link:detach-forced"
+    LinkTransferLimitExceeded = b"amqp:link:transfer-limit-exceeded"
+    LinkMessageSizeExceeded = b"amqp:link:message-size-exceeded"
+    # TODO: check whether Client/Unknown/Vendor Error are exposed in EH/SB as users might be depending
+    #  on the code for error handling
+    ClientError = b"amqp:client-error"
+    UnknownError = b"amqp:unknown-error"
+    VendorError = b"amqp:vendor-error"
+    SocketError = b"amqp:socket-error"
+
+
+class RetryMode(str, Enum):
+    EXPONENTIAL = 'exponential'
+    FIXED = 'fixed'
+
+
+class RetryPolicy:
+
+    no_retry = [
+        ErrorCodes.DecodeError,
+        ErrorCodes.LinkMessageSizeExceeded,
+        ErrorCodes.NotFound,
+        ErrorCodes.NotImplemented,
+        ErrorCodes.LinkRedirect,
+        ErrorCodes.NotAllowed,
+        ErrorCodes.UnauthorizedAccess,
+        ErrorCodes.LinkStolen,
+        ErrorCodes.ResourceLimitExceeded,
+        ErrorCodes.ConnectionRedirect,
+        ErrorCodes.PreconditionFailed,
+        ErrorCodes.InvalidField,
+        ErrorCodes.ResourceDeleted,
+        ErrorCodes.IllegalState,
+        ErrorCodes.FrameSizeTooSmall,
+        ErrorCodes.ConnectionFramingError,
+        ErrorCodes.SessionUnattachedHandle,
+        ErrorCodes.SessionHandleInUse,
+        ErrorCodes.SessionErrantLink,
+        ErrorCodes.SessionWindowViolation
+    ]
+
+    def __init__(
+        self,
+        **kwargs
+    ):
+        """
+        keyword int retry_total:
+        keyword float retry_backoff_factor:
+        keyword float retry_backoff_max:
+        keyword RetryMode retry_mode:
+        keyword list no_retry_condition:
+        keyword dict custom_retry_policy:
+        """
+        self.total_retries = kwargs.pop('retry_total', 3)
+        # TODO: A. consider letting retry_backoff_factor be either a float or a callback obj which returns a float
+        #  to give more extensibility on customization of retry backoff time, the callback could take the exception
+        #  as input.
+        self.backoff_factor = kwargs.pop('retry_backoff_factor', 0.8)
+        self.backoff_max = kwargs.pop('retry_backoff_max', 120)
+        self.retry_mode = kwargs.pop('retry_mode', RetryMode.Exponential)
+        self.custom_no_retry = self.no_retry + kwargs.pop("no_retry_condition", [])
```

Comment:
Rather than make a `custom_no_retry` list, we can just have a parameter called `no_retry`, then:
```
self.no_retry.extend(kwargs.get('no_retry', []))
```

Reply:
nice suggestion, updated!